pkg/crossplane/instance.go

```diff
@@ -151,3 +155,58 @@ func (cp *Crossplane) GetInstance(ctx context.Context, instanceID string) (*comp
 	}
 	return nil, ErrInstanceNotFound
 }
+
+// UpdateInstanceSLA updates the SLA of an instance specified by the supplied planID.
+// Only SLA changes are allowed, any other change is not permitted and yields an error.
+func (cp *Crossplane) UpdateInstanceSLA(ctx context.Context, instanceID, serviceID, planID string) error {
+	instance, err := cp.GetInstance(ctx, instanceID)
+	if err != nil {
+		return err
+	}
+
+	instanceLabels := instance.GetLabels()
+	if serviceID != instanceLabels[ServiceIDLabel] {
+		return ErrServiceUpdateNotPermitted
+	}
+
+	newPlan, err := cp.GetPlan(ctx, planID)
+	if err != nil {
+		return err
+	}
+
+	instanceSLA := instanceLabels[SLALabel]
+	newPlanSLA := newPlan.Labels[SLALabel]
+
+	slaChangePermitted := func() bool {
+		if instanceSLA == SLAPremium && newPlanSLA == SLAStandard {
+			return true
+		}
+		if instanceSLA == SLAStandard && newPlanSLA == SLAPremium {
+			return true
+		}
+		return false
+	}
+
+	if !slaChangePermitted() {
+		return ErrSLAChangeNotPermitted
+	}
+
+	instance.SetCompositionReference(&corev1.ObjectReference{
+		Name: newPlan.Name,
+	})
+	for _, l := range []string{
+		PlanNameLabel,
+		SLALabel,
+	} {
+		instanceLabels[l] = newPlan.Labels[l]
+	}
+	instance.SetLabels(instanceLabels)
+
+	gvk, err := gvkFromPlan(newPlan)
+	if err != nil {
+		return err
+	}
+	instance.SetGroupVersionKind(gvk)
```

Comment:
This shouldn't be necessary since it's already set, no?

Reply:
I thought as well but it's somehow lost - I have to dig deeper to check why.

Reply:
https://github.com/kubernetes/apimachinery/blob/master/pkg/runtime/helper.go#L250-L259

This decoder is used and removes GVK apparently. Not sure what the reasoning behind is, TBH.